src/coprocessor/dag/batch/executors/util/aggr_executor.rs

```diff
@@ -284,16 +285,19 @@ impl<Src: BatchExecutor, I: AggregationExecutorImpl<Src>> BatchExecutor
                 // When there are error, we can just return empty data.
                 self.is_ended = true;
                 BatchExecuteResult {
-                    data: LazyBatchColumnVec::empty(),
+                    physical_columns: LazyBatchColumnVec::empty(),
+                    logical_rows: Vec::new(),
                     warnings: self.entities.context.take_warnings(),
                     is_drained: Err(e),
                 }
             }
             Ok((data, src_is_drained)) => {
                 self.is_ended = src_is_drained;
-                let data = data.unwrap_or_else(LazyBatchColumnVec::empty);
+                let logical_columns = data.unwrap_or_else(LazyBatchColumnVec::empty);
+                let logical_rows = (0..logical_columns.rows_len()).collect();
```

Comment:
A `LazyBatchColumnVec` doesn't hold the assumption that all rows in it are valid, although it is true in aggregation's `handle_next_batch`. So this line makes me uncomfortable. Maybe we'd better assert it in some ways...
In fact, I do hope there is a type specifically for vectors whose logical rows are just all rows...They are common and there is overhead on processing them now (for example the `Arc[usize]`).

Reply:
Yes. `LazyBatchColumnVec` does not hold the assumption but `aggregate_partial_results` holds.

No special types are introduced for the case when logical rows are all rows (let's call it compact) mostly because we cannot determine it at compile time. For example, since we don't know what the underlying executor of the aggregation executor is, we cannot know whether or not they are compact at compile time. To resolve it, we need branches, or virtual dispatches anyway, which seems to have similar performance as a required `logical_rows` array.